src/core/net/dnssd_server.cpp

```diff
@@ -107,13 +115,21 @@ void Server::HandleUdpReceive(void *aContext, otMessage *aMessage, const otMessa
 
 void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
 {
-    Error  error = kErrorNone;
     Header requestHeader;
 
-    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), requestHeader));
-    VerifyOrExit(requestHeader.GetType() == Header::kTypeQuery, error = kErrorDrop);
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE
+    // We first let the `Srp::Server` process the received message.
+    // It returns `kErrorNone` to indicate that it successfully
```

Comment:
I would prefer to dispatch UDP messages according to the DNS Header:
- Dispatch to SRP server if DNS type is kTypeQuery and Query Type is kQueryTypeUpdate
- DNS-SD server process other messages

I think it's not safe to rely on errors.
Thoughts? @abtink @wgtdkp

Reply:
If you look into the implementation of `HandleDnssdServerUdpReceive`, it's exactly dispatching the messages with DNS type.

Reply:
We need the dispatching/filtering in `Srp::Server` independent of whether we share the socket or use our own socket. Even if we use our own socket we still may receive other type of messages on the socket and need to filter (accept DNS type query `kQueryTypeUpdate` only).

We can add the dispatching/filtering code in `HandleDnssdServerUdpReceive()` but then we would need to repeat the same code when we use our own socket.

So it make sense to have a common place where we do this (i.e., in `ProcessMessage()`) and use it from both callbacks (whether we receive from our own socket `HandleUdpReceive()` or from shared socket `HandleDnssdServerUdpReceive()`).

Reply:
SGTM